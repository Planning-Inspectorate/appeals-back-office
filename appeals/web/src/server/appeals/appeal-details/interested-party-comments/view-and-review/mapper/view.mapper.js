import { appealShortReference } from '#lib/appeals-formatter.js';
import { generateCommentSummaryList, generateWithdrawLink } from './common.js';

/** @typedef {import("../../../appeal-details.types.js").WebAppeal} Appeal */
/** @typedef {import("../../interested-party-comments.types.js").Representation} Representation */

/**
 * @param {Appeal} appealDetails
 * @param {Representation} comment
 * @returns {PageContent}
 */
export function viewInterestedPartyCommentPage(appealDetails, comment) {
	const shortReference = appealShortReference(appealDetails.appealReference);
	const commentSummaryList = generateCommentSummaryList(comment);
	const withdrawLink = generateWithdrawLink();

	const pageContent = {
		title: 'View comment',
		backLinkUrl: `/appeals-service/appeal-details/${appealDetails.appealId}/interested-party-comments`,
		preHeading: `Appeal ${shortReference}`,
		heading: 'View comment',
		headingClasses: 'govuk-heading-l',
		pageComponents: [commentSummaryList, withdrawLink]
	};

	return pageContent;
}
