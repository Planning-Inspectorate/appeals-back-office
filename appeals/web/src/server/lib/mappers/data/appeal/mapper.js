import usersService from '#appeals/appeal-users/users-service.js';
import config from '#environment/config.js';
import { permissionNames } from '#environment/permissions.js';
import { userHasPermission } from '#lib/mappers/index.js';
import { APPEAL_TYPE } from '@pins/appeals/constants/common.js';
import { submaps as s78Submaps } from './s78.js';
import { submaps as hasSubmaps } from './has.js';

/**
 * @typedef {import('#appeals/appeal-details/appeal-details.types.js').WebAppeal} WebAppeal
 * @typedef {import('../../../../app/auth/auth-session.service.js').SessionWithAuth} SessionWithAuth
 */

/**
 * @typedef {Object} SubMapperParams
 * @property {WebAppeal} appealDetails
 * @property {string} currentRoute
 * @property {SessionWithAuth} session
 * @property {boolean} userHasUpdateCasePermission
 * @property {boolean} skipAssignedUsersData
 * @property {import('@pins/express/types/express.js').Request} request
 * @property {{
 *     [x: string]: any;
 *   } | null} [caseOfficerUser]
 * @property {{
 *     [x: string]: any;
 *   } | null} [inspectorUser]
 * @property {import('#appeals/appeal-details/representations/representations.service.js').Representation} [appellantFinalComments]
 * @property {import('#appeals/appeal-details/representations/representations.service.js').Representation} [lpaFinalComments]
 * @property {import('@pins/appeals.api').Appeals.SingleAppellantCaseResponse} [appellantCase]
 */

/**
 * @typedef {(params: SubMapperParams) => Instructions} SubMapper
 */

/** @type {Record<string, Record<string, SubMapper>>} */
const submaps = {
	[APPEAL_TYPE.HOUSEHOLDER]: hasSubmaps,
	[APPEAL_TYPE.S78]: s78Submaps,
	[APPEAL_TYPE.PLANNED_LISTED_BUILDING]: s78Submaps
};

/**
 * @param {WebAppeal} appealDetails
 * @param {string} currentRoute
 * @param {SessionWithAuth} session
 * @param {import('@pins/express/types/express.js').Request} request
 * @param {boolean} [skipAssignedUsersData]
 * @param {import('#appeals/appeal-details/representations/representations.service.js').Representation} [appellantFinalComments]
 * @param {import('#appeals/appeal-details/representations/representations.service.js').Representation} [lpaFinalComments]
 * @param {import('@pins/appeals.api').Appeals.SingleAppellantCaseResponse} [appellantCase]
 * @returns {Promise<{appeal: MappedInstructions}>}
 */
export async function initialiseAndMapAppealData(
	appealDetails,
	currentRoute,
	session,
	request,
	skipAssignedUsersData = false,
	appellantFinalComments,
	lpaFinalComments,
	appellantCase
) {
	if (appealDetails === undefined) {
		throw new Error('appealDetails is undefined');
	}

	if (!appealDetails.appealType) {
		throw new Error('No appealType on appealDetails');
	}

	const caseOfficerUser = appealDetails.caseOfficer
		? await usersService.getUserByRoleAndId(
				config.referenceData.appeals.caseOfficerGroupId,
				session,
				appealDetails.caseOfficer
		  )
		: null;

	const inspectorUser = appealDetails.inspector
		? await usersService.getUserByRoleAndId(
				config.referenceData.appeals.inspectorGroupId,
				session,
				appealDetails.inspector
		  )
		: null;

	const userHasUpdateCasePermission = userHasPermission(permissionNames.updateCase, session);

	/** @type {Record<string, SubMapper>} */
	const submappers = submaps[appealDetails.appealType];

	/** @type {{appeal: MappedInstructions}} */
	const mappedData = {
		appeal: {}
	};

	Object.entries(submappers).forEach(([key, submapper]) => {
		mappedData.appeal[key] = submapper({
			appealDetails,
			currentRoute,
			session,
			request,
			skipAssignedUsersData,
			userHasUpdateCasePermission,
			caseOfficerUser,
			inspectorUser,
			appellantFinalComments,
			lpaFinalComments,
			appellantCase
		});
	});
	return mappedData;
}
