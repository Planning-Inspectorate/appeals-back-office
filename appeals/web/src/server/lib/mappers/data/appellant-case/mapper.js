import { permissionNames } from '#environment/permissions.js';
import { APPEAL_TYPE } from '@pins/appeals/constants/common.js';
import { userHasPermission } from '#lib/mappers/index.js';
import { hasRowMaps } from './has.js';
import { s78RowMaps } from './s78.js';

/**
 * @typedef RowMapperParams
 * @property {import('@pins/appeals.api').Appeals.SingleAppellantCaseResponse} appellantCaseData
 * @property {import('#appeals/appeal-details/appeal-details.types.js').WebAppeal} appealDetails
 * @property {string} currentRoute
 * @property {boolean} userHasUpdateCase
 */

/**
 * @typedef {(params: RowMapperParams) => Instructions} RowMapper
 */

/** @type {Record<string, Record<string, RowMapper>>} */
const rowMaps = {
	[APPEAL_TYPE.D]: hasRowMaps,
	[APPEAL_TYPE.W]: s78RowMaps
};

/**
 * @param {import('@pins/appeals.api').Appeals.SingleAppellantCaseResponse} appellantCaseData
 * @param {import('#appeals/appeal-details/appeal-details.types.js').WebAppeal} appealDetails
 * @param {import('../../../../app/auth/auth-session.service.js').SessionWithAuth} session
 * @param {string} currentRoute
 * @returns {MappedInstructions}
 */
export function initialiseAndMapData(appellantCaseData, appealDetails, currentRoute, session) {
	if (!appellantCaseData || appellantCaseData === null) {
		throw new Error('appellantCaseDetails is null or undefined');
	}
	if (appealDetails === undefined) {
		throw new Error('appealDetails is undefined');
	}

	if (!appealDetails.appealType) {
		throw new Error('No appealType on appealDetails');
	}

	currentRoute =
		currentRoute[currentRoute.length - 1] === '/' ? currentRoute.slice(0, -1) : currentRoute;

	const userHasUpdateCase = userHasPermission(permissionNames.updateCase, session);

	const rowMapperParams = {
		appellantCaseData,
		appealDetails,
		currentRoute,
		userHasUpdateCase
	};
	/** @type {Record<string, RowMapper>} */
	const rowMappers = rowMaps[appealDetails.appealType];

	/** @type {MappedInstructions} */
	const mappedData = {};

	Object.entries(rowMappers).forEach(([key, rowMapper]) => {
		mappedData[key] = rowMapper(rowMapperParams);
	});

	return mappedData;
}
