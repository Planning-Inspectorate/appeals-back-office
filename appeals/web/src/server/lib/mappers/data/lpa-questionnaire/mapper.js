import { permissionNames } from '#environment/permissions.js';
import { userHasPermission } from '#lib/mappers/index.js';
import { APPEAL_TYPE } from '@pins/appeals/constants/common.js';
import { submaps as casAdvertsSubMaps } from './cas-advert.js';
import { submaps as casPlanningSubMaps } from './cas-planning.js';
import { submaps as hasSubmaps } from './has.js';
import { submaps as s20Submaps } from './s20.js';
import { submaps as s78Submaps } from './s78.js';
/**
 * @typedef {import('#appeals/appeal-details/appeal-details.types.js').WebAppeal} WebAppeal
 * @typedef {import('../../../../app/auth/auth-session.service.js').SessionWithAuth & Express.Request["session"]} SessionWithAuth
 * @typedef {import('@pins/appeals.api').Appeals.SingleLPAQuestionnaireResponse} SingleLPAQuestionnaireResponse
 */

/**
 * @typedef {Object} SubMapperParams
 * @property {SingleLPAQuestionnaireResponse} lpaQuestionnaireData
 * @property {WebAppeal} appealDetails
 * @property {string} currentRoute
 * @property {SessionWithAuth} session
 * @property {boolean} userHasUpdateCase
 */

/**
 * @typedef {(params: SubMapperParams) => Instructions} SubMapper
 */
/**
/** @type {Record<string, Record<string, SubMapper>>} */
const submaps = {
	[APPEAL_TYPE.HOUSEHOLDER]: hasSubmaps,
	[APPEAL_TYPE.CAS_ADVERTISEMENT]: casAdvertsSubMaps,
	[APPEAL_TYPE.ADVERTISEMENT]: casAdvertsSubMaps,
	[APPEAL_TYPE.CAS_PLANNING]: casPlanningSubMaps,
	[APPEAL_TYPE.S78]: s78Submaps,
	[APPEAL_TYPE.PLANNED_LISTED_BUILDING]: s20Submaps
};

/**
 * @param {SingleLPAQuestionnaireResponse} lpaQuestionnaireData
 * @param {WebAppeal} appealDetails
 * @param {string} currentRoute
 * @param {SessionWithAuth} session
 * @returns {{lpaq: MappedInstructions}}
 */
export function initialiseAndMapLPAQData(
	lpaQuestionnaireData,
	appealDetails,
	session,
	currentRoute
) {
	if (appealDetails === undefined) {
		throw new Error('appealDetails is undefined');
	}

	if (!appealDetails.appealType) {
		throw new Error('No appealType on appealDetails');
	}
	const userHasUpdateCase = userHasPermission(permissionNames.updateCase, session);

	/** @type {{lpaq: MappedInstructions}} */
	const mappedData = { lpaq: {} };

	const submappers = submaps[appealDetails.appealType];

	Object.entries(submappers).forEach(([key, submapper]) => {
		mappedData.lpaq[key] = submapper({
			lpaQuestionnaireData,
			appealDetails,
			currentRoute,
			session,
			userHasUpdateCase
		});
	});

	return mappedData;
}
