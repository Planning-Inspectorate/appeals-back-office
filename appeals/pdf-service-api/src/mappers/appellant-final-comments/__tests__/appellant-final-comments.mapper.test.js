import { mapAppellantFinalComments } from '../appellant-final-comments.mapper.js';
import mockAppellantFinalCommentsData from '../../../mocks/mock-appellant-final-comments-data.json';

describe('mapAppellantFinalCommentsData', () => {
	it('should map appellant final comments data correctly', () => {
		const result = mapAppellantFinalComments(
			mockAppellantFinalCommentsData.appellantFinalCommentsData
		);
		expect(result).toMatchSnapshot();
	});
});
