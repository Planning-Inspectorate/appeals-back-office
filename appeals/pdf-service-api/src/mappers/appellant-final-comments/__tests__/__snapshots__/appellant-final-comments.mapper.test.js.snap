// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapAppellantFinalCommentsData should map appellant final comments data correctly 1`] = `
{
  "dateReceived": "3 July 2025, 16:14",
  "details": {
    "agent": {
      "email": "test1@example.com",
      "firstName": "Lee",
      "lastName": "Thornton",
      "organisationName": "Lee Thornton Ltd",
      "phoneNumber": null,
      "serviceUserId": 105,
    },
    "appealId": 72,
    "appealReference": "6000072",
    "appealSite": {
      "addressId": 72,
      "addressLine1": "Copthalls",
      "addressLine2": "Clevedon Road",
      "postCode": "BS48 1PN",
      "town": "West Hill",
    },
    "appealStatus": "final_comments",
    "appealTimetable": {
      "appealTimetableId": 37,
      "appellantStatementDueDate": "2025-02-07T23:59:00.000Z",
      "caseResubmissionDueDate": null,
      "finalCommentsDueDate": "2025-02-21T23:59:00.000Z",
      "ipCommentsDueDate": "2025-02-07T23:59:00.000Z",
      "lpaQuestionnaireDueDate": "2025-01-10T23:59:00.000Z",
      "lpaStatementDueDate": "2025-02-07T23:59:00.000Z",
      "planningObligationDueDate": null,
      "s106ObligationDueDate": "2025-02-21T23:59:00.000Z",
      "statementOfCommonGroundDueDate": null,
    },
    "appealType": "Planning appeal",
    "appellant": {
      "email": "test4@example.com",
      "firstName": "Sophie",
      "lastName": "Skinner",
      "organisationName": null,
      "phoneNumber": null,
      "serviceUserId": 106,
    },
    "appellantCaseId": 72,
    "attachments": [],
    "author": "Sophie Skinner",
    "completedStateList": [],
    "costs": {
      "appellantApplicationFolder": {
        "caseId": 72,
        "documents": [],
        "folderId": 3587,
        "path": "costs/appellantCostsApplication",
      },
      "appellantCorrespondenceFolder": {
        "caseId": 72,
        "documents": [],
        "folderId": 3589,
        "path": "costs/appellantCostsCorrespondence",
      },
      "appellantDecisionFolder": {
        "caseId": 72,
        "documents": [],
        "folderId": 3593,
        "path": "costs/appellantCostsDecisionLetter",
      },
      "appellantWithdrawalFolder": {
        "caseId": 72,
        "documents": [],
        "folderId": 3588,
        "path": "costs/appellantCostsWithdrawal",
      },
      "lpaApplicationFolder": {
        "caseId": 72,
        "documents": [],
        "folderId": 3590,
        "path": "costs/lpaCostsApplication",
      },
      "lpaCorrespondenceFolder": {
        "caseId": 72,
        "documents": [],
        "folderId": 3592,
        "path": "costs/lpaCostsCorrespondence",
      },
      "lpaDecisionFolder": {
        "caseId": 72,
        "documents": [],
        "folderId": 3594,
        "path": "costs/lpaCostsDecisionLetter",
      },
      "lpaWithdrawalFolder": {
        "caseId": 72,
        "documents": [],
        "folderId": 3591,
        "path": "costs/lpaCostsWithdrawal",
      },
    },
    "created": "2025-07-03T15:14:00.044Z",
    "createdAt": "2025-07-03T15:13:55.967Z",
    "decision": {
      "folderId": 3599,
      "path": "appeal-decision/caseDecisionLetter",
    },
    "documentationSummary": {
      "appellantCase": {
        "dueDate": null,
        "receivedAt": "2025-07-03T15:13:55.967Z",
        "status": "Valid",
      },
      "appellantFinalComments": {
        "receivedAt": "2025-07-03T15:14:00.044Z",
        "representationStatus": "awaiting_review",
        "status": "received",
      },
      "ipComments": {
        "counts": {
          "awaiting_review": 5,
          "invalid": 4,
          "valid": 1,
        },
        "isRedacted": false,
        "receivedAt": "2025-07-03T15:14:00.000Z",
        "status": "received",
      },
      "lpaFinalComments": {
        "receivedAt": "2025-07-03T15:14:00.065Z",
        "representationStatus": "awaiting_review",
        "status": "received",
      },
      "lpaQuestionnaire": {
        "dueDate": "2025-01-10T23:59:00.000Z",
        "receivedAt": "2023-05-08T23:00:00.000Z",
        "status": "received",
      },
      "lpaStatement": {
        "isRedacted": false,
        "receivedAt": "2025-07-03T15:14:00.031Z",
        "representationStatus": "valid",
        "status": "received",
      },
    },
    "eiaScreeningRequired": null,
    "environmentalAssessment": {
      "caseId": 72,
      "documents": [],
      "folderId": 3563,
      "path": "appellant-case/environmentalAssessment",
    },
    "healthAndSafety": {
      "appellantCase": {
        "details": null,
        "hasIssues": false,
      },
      "lpaQuestionnaire": {
        "details": "There may be no mobile reception at the site",
        "hasIssues": true,
      },
    },
    "id": 239,
    "inspectorAccess": {
      "appellantCase": {
        "details": null,
        "isRequired": false,
      },
      "lpaQuestionnaire": {
        "details": "There is a tall hedge around the site which obstructs the view of the site",
        "isRequired": true,
      },
    },
    "internalCorrespondence": {
      "crossTeam": {
        "caseId": 72,
        "documents": [],
        "folderId": 3595,
        "path": "internal/crossTeamCorrespondence",
      },
      "inspector": {
        "caseId": 72,
        "documents": [],
        "folderId": 3596,
        "path": "internal/inspectorCorrespondence",
      },
      "mainParty": {
        "caseId": 72,
        "documents": [],
        "folderId": 3597,
        "path": "internal/mainPartyCorrespondence",
      },
    },
    "isChildAppeal": false,
    "isParentAppeal": false,
    "linkedAppeals": [],
    "localPlanningDepartment": "Maidstone Borough Council",
    "lpaEmailAddress": "maid@lpa-email.gov.uk",
    "lpaQuestionnaireId": 40,
    "neighbouringSites": [],
    "notes": "",
    "origin": "citizen",
    "originalRepresentation": "Final comment from appellant. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore.",
    "otherAppeals": [],
    "planningApplicationReference": "44377/APP/3/218453",
    "procedureType": "Written",
    "redactedRepresentation": "",
    "rejectionReasons": [],
    "representationType": "appellant_final_comment",
    "represented": {
      "address": {
        "addressLine1": "",
        "postCode": "",
      },
      "email": "test4@example.com",
      "id": 106,
      "name": "Sophie Skinner",
    },
    "siteVisitRequested": false,
    "source": "citizen",
    "startedAt": "2025-01-03T16:13:52.867Z",
    "stateList": [
      {
        "completed": true,
        "key": "assign_case_officer",
      },
      {
        "completed": true,
        "key": "validation",
      },
      {
        "completed": true,
        "key": "ready_to_start",
      },
      {
        "completed": true,
        "key": "lpa_questionnaire",
      },
      {
        "completed": true,
        "key": "statements",
      },
      {
        "completed": false,
        "key": "final_comments",
      },
      {
        "completed": false,
        "key": "event",
      },
      {
        "completed": false,
        "key": "evidence",
      },
      {
        "completed": false,
        "key": "awaiting_event",
      },
      {
        "completed": false,
        "key": "issue_determination",
      },
      {
        "completed": false,
        "key": "awaiting_transfer",
      },
      {
        "completed": false,
        "key": "invalid",
      },
      {
        "completed": false,
        "key": "transferred",
      },
      {
        "completed": false,
        "key": "closed",
      },
      {
        "completed": false,
        "key": "withdrawn",
      },
      {
        "completed": false,
        "key": "complete",
      },
    ],
    "status": "awaiting_review",
    "validAt": "2024-09-03T15:13:52.867Z",
    "withdrawal": {
      "withdrawalFolder": {
        "caseId": 72,
        "documents": [],
        "folderId": 3555,
        "path": "appellant-case/appellantCaseWithdrawalLetter",
      },
    },
  },
  "heading": "Appellant final comments",
  "sections": [
    {
      "details": String {
        "length": 1035,
        "val": "Final comment from appellant. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore.",
      },
      "heading": "Submitted final comments",
    },
    {
      "details": "No documents",
      "heading": "Supporting documents",
    },
  ],
  "status": "Awaiting review",
}
`;
